app/views/boxes/_form.haml

```diff
@@ -43,3 +43,64 @@
   = f.form_actions do
     = f.submit 'Save', class: 'btn-primary', id: 'btn-save'
     = link_to 'Cancel', boxes_path, class: 'btn-link'
+
+.sample-source-change-modal-container.hidden
+  = react_component("ConfirmationModal", 
+                      deletion: true,
+                      cancelFunction: "cancelSampleSourceChange",
+                      confirmFunction: "confirmSampleSourceChange", 
+                      id: "sample-source-confirmation-modal", 
+                      colorClass: "red", 
+                      confirmMessage: "Continue", 
+                      title: "Warning", 
+                      message: "If you change the sample source, the box contents will be emptied. ")
+
+:javascript
+  
+  var confirmSampleSourceChange = function() {}
+  var cancelSampleSourceChange = function() {}
+
+  var sourcesRadio = document.querySelectorAll(".radiotoggle input");
+  var oldSource = null;
+
+  var applySampleSourceChange = function (displayField) {
+    // Hide all the fields and disable them
+    sourcesRadio.forEach(function(otherRadio) {
+      document.querySelector(`#${otherRadio.value}`).style.display = 'none';
+      document.querySelector(`#${otherRadio.value}`).setAttribute("disabled", true);
+    });
+    // Show the selected field and enable it
+    document.querySelector(`#${displayField.value}`).style.display = 'block';
+    document.querySelector(`#${displayField.value}`).removeAttribute("disabled");
+    // Clear the selectors contents
+    document.querySelector(".clear-batches").click()
+    document.querySelector(".clear-samples").click()
+  }
+
+  sourcesRadio.forEach(function(radio) {
+    radio.addEventListener("change", function(evt) {
+      var thisField = this;
+
+      confirmSampleSourceChange = function() {
+        applySampleSourceChange(thisField);
+        document.querySelector(`.sample-source-change-modal-container`).style.visibility = 'hidden';
+        // Update the old source only if the user confirmed the change
+        oldSource = thisField;
+      }
+      
+      cancelSampleSourceChange = function() {
+        document.querySelector(`.sample-source-change-modal-container`).style.visibility = 'hidden';
+        // Check back the old source since it was not changed
+        oldSource.checked = true;
+      }
+
+      if (oldSource) {
+        // If the user changed the source, show the modal
+        document.querySelector(`.sample-source-change-modal-container`).style.visibility = 'visible';
+      } else {
+        // For the first time, simulate that the user clicked on the confirm button
+        confirmSampleSourceChange();
+      }
+    });
+  });
```

Comment:
I think we can use global functions. We can take advantage that the functions will only ever be invoked _after_ showing the confirm modal (or faking a click), so we don't need them to be noop by default.

```js
function confirmSampleSourceChange() {
  applySampleSourceChange(this);
  // ...
}

function cancelSampleSourceChange() {
  // ...
}

sourcesRadio.forEach(function(radio) {
  radio.addEventListener("change", function (evt) {
    if (oldSource) {
      // ..
    } else {
      // here is the trick to have a specific `this`:
      confirmSampleSourceChange.call(this)
    }
  });
});
```

Reply:
This won't work as it is: remember that `confirmSampleSourceChange` is also passed to the custom confirm modal as a function to be called when the confirm button is pressed, and there the `this` parameter will be the window. I need to redefine the `confirmSampleSourceChange` function (as a var) each time the `change` event is triggered.

Reply:
This won't work as it is: remember that `confirmSampleSourceChange` is also passed to the custom confirm modal as a function to be called when the confirm button is pressed, and there the `this` parameter will be the window. I need to redefine the `confirmSampleSourceChange` function (as a var) each time the `change` event is triggered.

Reply:
Oh right :thinking: Then what about the following?

```js
var confirmSampleSourceChange;
function confirmSampleSourceChangeFn() {}
function cancelSampleSourceChange() {}

sourcesRadio.forEach(function(radio) {
  radio.addEventListener("change", function (evt) {
    if (oldSource) {
      confirmSampleSourceChange = confirmSampleSourceChangeFn.bind(this)
      // ...
    } else {
      confirmSampleSourceChangeFn.call(this)
    }
  });
});
```

We could also have `showConfirmModal()` and `hideConfirmModal()` helpers to improve readability.